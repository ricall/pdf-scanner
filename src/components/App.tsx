import React, { useState } from 'react';
import { Box } from "@material-ui/core";
import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import { usePersistState } from 'persist-state';
import { initialState, PdfDocument, PdfMetadata } from "../types";
import { pdfReducer, useAsyncReducer } from "../reducers";
import {
    deletePage,
    loadDocument,
    movePageLeft,
    movePageRight,
    newDocument,
    rotatePage,
    saveDocument,
    setCurrentPage,
    updateMetadata
} from "../actions";
import { ApplicationAppBar } from "./ApplicationAppBar";
import { defaultSettings, Settings, SettingsDrawer } from "./SettingsDrawer";
import { MetadataForm } from "./MetadataForm";
import { ConfirmDialog } from "./ConfirmDialog";
import { PageGrid } from "./PageGrid";

const useStyles = makeStyles((theme: Theme) => createStyles({
    mainBody: {
        minHeight: `80vh`,
    }
}));

const isDocumentValid = ({ metadata: { title, author } }: PdfDocument): boolean => !!title && !!author;

export const App = () => {
    const classes = useStyles();
    const [ settings, setSettings ] = usePersistState(defaultSettings(), {
        key: 'settings',
        persistAcrosSession: true,
        persistOnUnmount: true,
    });
    const [ state, dispatch ] = useAsyncReducer(pdfReducer, initialState);
    const [ editSettings, setEditSettings ] = useState(false);
    const [ confirmNewDocument, setConfirmNewDocument ] = useState(false);
    const [ editMetadata, setEditMetadata ] = useState(false);
    const { name, pages, metadata } = state;

    const onAddDocument = (name: string, document: ArrayBuffer) => dispatch(loadDocument(settings, name, document));
    const onSetCurrentPage = (page: number) => dispatch(setCurrentPage(page));
    const onRotatePage = (page: number) => dispatch(rotatePage(page));
    const onMoveLeft = (page: number) => dispatch(movePageLeft(page));
    const onMoveRight = (page: number) => dispatch(movePageRight(page));
    const onDelete = (page: number) => dispatch(deletePage(page));
    const onSaveDocument = () => {
        if (isDocumentValid(state)) {
            dispatch(saveDocument(settings, state));
        } else {
            setEditMetadata(true);
        }
    }

    // Menu Settings
    const onMenu = () => setEditSettings(true);
    const onSettingsChange = (settings: Settings) => {
        setEditSettings(false);
        setSettings(settings);
    };

    // Confirm New Document
    const onConfirmNewDocument = () => {
        setConfirmNewDocument(false);
        dispatch(newDocument());
    }
    const onRejectNewDocument = () => setConfirmNewDocument(false);
    const onNewDocument = () => {
        if (pages.length > 0) {
            setConfirmNewDocument(true);
        } else {
            onConfirmNewDocument();
        }
    }

    // Document Metadata
    const onEditMetadata = () => setEditMetadata(true);
    const onMetadataChange = async (metadata: PdfMetadata, accept: boolean) => {
        setEditMetadata(false);
        if (accept) {
            dispatch(updateMetadata(metadata));
        }
    }

    return (
        <div className="App">
            <ApplicationAppBar
                title={name}
                pages={pages.length}
                onMenu={onMenu}
                onNewDocument={onNewDocument}
                onSaveDocument={onSaveDocument}
                onEditMetadata={onEditMetadata}
            />
            <Box className={classes.mainBody} display="flex" justifyContent="center" alignItems="center">
                <PageGrid
                    document={state}
                    setCurrentPage={onSetCurrentPage}
                    onRotate={onRotatePage}
                    onMoveLeft={onMoveLeft}
                    onMoveRight={onMoveRight}
                    onDelete={onDelete}
                    onAddDocument={onAddDocument}
                />
            </Box>
            {editSettings && <SettingsDrawer settings={settings} onChange={onSettingsChange}/>}
            {editMetadata && <MetadataForm metadata={metadata} onChange={onMetadataChange}/>}
            <ConfirmDialog display={confirmNewDocument} onAccept={onConfirmNewDocument} onReject={onRejectNewDocument}/>
        </div>
    );
}