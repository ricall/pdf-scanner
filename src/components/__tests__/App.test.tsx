import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { App } from '../App';

describe('App component', () => {

    it('can be rendered', () => {
        const wrapper = shallow(<App />);
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();
    });

});
