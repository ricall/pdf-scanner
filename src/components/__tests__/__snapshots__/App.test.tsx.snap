// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component can be rendered 1`] = `
<div
  className="App"
>
  <ApplicationAppBar
    onEditMetadata={[Function]}
    onMenu={[Function]}
    onNewDocument={[Function]}
    onSaveDocument={[Function]}
    pages={0}
    title="untitled"
  />
  <Styled(MuiBox)
    alignItems="center"
    className="makeStyles-mainBody-1"
    display="flex"
    justifyContent="center"
  >
    <PageGrid
      document={
        Object {
          "bytes": ArrayBuffer [],
          "currentPage": 0,
          "metadata": Object {
            "author": "",
            "creationDate": undefined,
            "creator": "",
            "keywords": "",
            "modificationDate": undefined,
            "producer": "",
            "subject": "",
            "title": "",
          },
          "name": "untitled",
          "pages": Array [],
        }
      }
      onAddDocument={[Function]}
      onDelete={[Function]}
      onMoveLeft={[Function]}
      onMoveRight={[Function]}
      onRotate={[Function]}
      setCurrentPage={[Function]}
    />
  </Styled(MuiBox)>
  <ConfirmDialog
    display={false}
    onAccept={[Function]}
    onReject={[Function]}
  />
</div>
`;
